runtime/Cpp/runtime/src/atn/PredictionContext.cpp

```diff
@@ -101,48 +100,46 @@ Ref<PredictionContext> PredictionContext::merge(const Ref<PredictionContext> &a,
     return a;
   }
 
-  if (is<SingletonPredictionContext>(a) && is<SingletonPredictionContext>(b)) {
-    return mergeSingletons(std::dynamic_pointer_cast<SingletonPredictionContext>(a),
-                           std::dynamic_pointer_cast<SingletonPredictionContext>(b), rootIsWildcard, mergeCache);
+  if (a->isSingleton() && b->isSingleton()) {
+    return mergeSingletons(std::static_pointer_cast<SingletonPredictionContext>(a),
+                           std::static_pointer_cast<SingletonPredictionContext>(b), rootIsWildcard, mergeCache);
   }
 
   // At least one of a or b is array.
   // If one is $ and rootIsWildcard, return $ as * wildcard.
   if (rootIsWildcard) {
-    if (is<EmptyPredictionContext>(a)) {
+    if (a->isEmpty()) {
```

Comment:
Do you have any evidence that using a virtual function is actually faster than a dynamic cast? Would be interesting to see some numbers which support this switch.

Btw, the name `isEmpty()` is not wisely choosen as it seems to indicate the context might be empty (regardless of it's type). It is probably clearer to call that `isEmptyContext` or similar, instead. Sounds like a tiny problem, but I'd like to keep the original semantic (the test is about a type check, not it's content). Not a strong required, however.

Reply:
isEmpty was a pre-existing method, but will change that as its ripe for miss-interpretation. I already at first missed the subtle point that an EmptyPredictionContext has a SingletonPredictionContext as a base class.
My evidence for the virtual function calls was that they were showing up in my VC++ profile runs. But if you know how dynamic_cast is implemented then the virtual function calls should be at least as fast as any dynamic_cast. Its possible that gcc/clang will figure out that the result of dynamic_cast is only being type checked against at high optimization levels but not on my release build profiles. I was going to use typeid() but thought this might cause subtle issues after any further changes to this class hierarchy.
Will try with/without and produce some numbers to support this change however.

Reply:
I compiled versions of the code with is<> and with isEmptyContext()/isSingletonContext() just for the code in PredictionContext.cpp. While I really noticed the is calls in ParserATNSimulator.cpp the timings indicate that the virtual function calls are an improvement here as well.
`
                             Virtual               dynamic_cast
Example 1(SLL)     806ms                840ms
                 (LL)      980ms                1019ms
Example 2(SLL)    2755ms               2826ms
                 (LL)     4951ms               4993ms
Example 3(SLL)    959ms                 1028ms
                 (LL)     20369ms             21239ms
Example 4 (SLL)   145ms                 152ms
                 (LL)    161919ms           166516ms
`
This is what I would expect, virtual function calls should be slightly faster than dynamic_cast in general.

Reply:
> Do you have any evidence that using a virtual function is actually faster than a dynamic cast? Would be interesting to see some numbers which support this switch.

It should always be no slower, and probably faster. A virtual call just looks up an entry in the vtable and calls it. A `dynamic_cast` has to do a lot more work to handle upcasts, downcasts, multiple inheritance hierarchies, etc. Even for a simple inheritance heirarchy, the C++ runtime still tends to implement dynamic cast using a complex algorithm that can handle the general case of complex hierarchies.
https://wyfeizj.wordpress.com/2011/03/25/the-rtti-implementation-of-vs2010/
https://tinodidriksen.com/2010/04/cpp-dynamic-cast-performance/ (I have no idea why this benchmark uses `reinterpret_cast` instead of `static_cast`, that is a smell, but doesn't affect the result in these specific tests).

A virtual function call is a lot easier to optimize too, as it's eligible for inlining via full devirtualization, or speculative devirt. http://hubicka.blogspot.com/2014/02/devirtualization-in-c-part-4-analyzing.html